scripts/yetong01_kinematic_trajectory_planning.cpp

```diff
@@ -0,0 +1,345 @@
+  }
+  for (size_t k = 1; k < num_steps; k++) {
+    Key curr_key = k;
+    Key next_key = k + 1;
+    Expression<SerialChain<7>> curr_state(curr_key);
+    Expression<SerialChain<7>> next_state(next_key);
+    for (const auto& joint : robot.joints()) {
+      auto joint_func = std::bind(&SerialChain<7>::joint, std::placeholders::_1,
+                                  joint->name(), std::placeholders::_2);
+      Expression<double> curr_joint(joint_func, curr_state);
+      Expression<double> next_joint(joint_func, next_state);
+      Expression<double> joint_rotation_expr = next_joint - curr_joint;
+      costs.addExpressionFactor<double>(joint_noise, 0.0, joint_rotation_expr);
+    }
+  }
+
+  // joint limit costs
+  JointLimitFunctor joint_limit_functor(joint_limit_low, joint_limit_high);
+  for (size_t k = 1; k <= num_steps; k++) {
+    Key curr_key = k;
+    Expression<SerialChain<7>> curr_state(curr_key);
+    for (const auto& joint : robot.joints()) {
+      auto joint_func = std::bind(&SerialChain<7>::joint, std::placeholders::_1,
+                                  joint->name(), std::placeholders::_2);
+      Expression<double> curr_joint(joint_func, curr_state);
+      Expression<double> joint_limit_error(joint_limit_functor, curr_joint);
+      costs.addExpressionFactor<double>(joint_limit_noise, 0.0,
+                                        joint_limit_error);
+    }
+  }
+
+  // target costs
+  Key last_key = num_steps;
+  Expression<SerialChain<7>> last_state(last_key);
+  auto pose_func = std::bind(&SerialChain<7>::link_pose, std::placeholders::_1,
+                             ee_name, std::placeholders::_2);
+  Expression<Pose3> target_pose_expr(pose_func, last_state);
+  costs.addExpressionFactor<Pose3>(target_noise, target_pose, target_pose_expr);
+  return costs;
+}
+
+/** Print joint angles for all steps. */
+void print_joint_angles(const Values& values) {
+  for (size_t k = 0; k <= num_steps; k++) {
+    std::cout << "step " << k << ":";
+    for (const auto& joint : robot.joints()) {
+      double angle = JointAngle(values, joint->id(), k);
+      std::cout << "\t" << angle;
+    }
+    std::cout << "\n";
+  }
+}
+
+/** Print joint angles of serial chain for all steps. */
+void print_joint_angles_sc(const Values& values) {
+  for (size_t k = 1; k <= num_steps; k++) {
+    Key state_key = k;
+    auto state = values.at<SerialChain<7>>(state_key);
+    std::cout << "step " << k << ":";
+    for (const auto& joint : robot.joints()) {
+      double angle = state.joint(joint->name());
+      std::cout << "\t" << angle;
+    }
+    std::cout << "\n";
+  }
+}
+
+/** Compare simple kinematic planning tasks of a robot arm using (1) dynamics
+ * factor graph (2) constraint manifold (3) manually specifed serial chain
+ * manifold. */
+void kinematic_planning() {
+  // problem
+  robot.fixLink(base_name);
+  auto constraints_graph = get_constraints_graph();
+  auto costs = get_costs();
```

Comment:
I don't understand this paradigm. If both `constraints_graph` and `costs` are factor graphs, why not just do something like this? I.e. have a single factor graph, but each function adds the necessary components to it.

```cpp
fg = get_constraints_graph();
fg = get_costs(fg);

optimize(fg, values);
```

Reply:
Good question! The constraints should be in the form of EqualityConstraints rather than factor graphs. The current constraints_graph represents the factors of merit functions for the constraints. Will revise to make it more clear.

Reply:
Done